[11.x] Add `Str::extract`

stevebauman opened the pull request:
### Description

This PR adds a `Str::extract` method to be able to retrieve pieces of text using a template string.

On various projects I've had to use different regex patterns to extract pieces of information. Such as ID's from URL's, information from comments, area codes from phone numbers, etc. A built-in way to extract this kind of data from known formats would have been a big time saver, and also help those who don't want to dive into their own regex to be able to do so.

### Usage

```php
use Illuminate\Support\Str;

// ['last_4' => '5000']
Str::extract(
    '4242-4242-4242-5000', 
    '*-*-*-{last_4}'
);

// ['area_code' => '800']
Str::extract(
    'Phone Number: 1-(800)-555-5555', 
    '*1-({area_code})-*'
);

// ['email' => 'john.doe@example.com']
Str::extract(
    'Contact us at john.doe@example.com.', 
    '* at {email}.'
);

// ['user_id' => '1', 'post_id' => '2']
Str::extract(
    '/users/1/posts/2/comments', 
    '/users/{user_id}/posts/{post_id}/*'
);

// ['day' => '1st', 'month' => 'January']
Str::extract(
    'My birthday is on the 1st of January', 
    'My birthday is on the {day} of {month}'
);

// ['street' => '123 Main St', 'city' => 'Anytown', 'region' => 'CA']
Str::extract(
    'The address is 123 Main St, Anytown, CA.', 
    '*address is {street}, {city}, {region}.'
);
```

stevebauman committed:
Add Str::extract

stevebauman committed:
Ensure strict matching so wildcards must be used

stevebauman committed:
Add tests

stevebauman committed:
Add to stringable

stevebauman committed:
Tweak test

stevebauman committed:
Spacing

stevebauman committed:
Add another test

stevebauman committed:
Spacing

stevebauman committed:
Use consistent quotations

stevebauman committed:
CS fixes

stevebauman committed:
More CS fixes

stevebauman committed:
Tweak comment

MrPunyapal commented on src/Illuminate/Support/Stringable.php:
```diff
@@ -295,6 +295,17 @@ public function excerpt($phrase = '', $options = [])
+     */
+    public function extract($pattern)
+    {
+        return Str::extract($pattern, $this->value);
```
Good catch 👏

stevebauman replied:
Whoopsies! Thanks @MeiKatz! 🙏 Fixed 👍

MeiKatz replied:
You're welcome

stevebauman committed:
Flip parameters

stevebauman committed:
Add sanity test

taylorotwell commented:
@stevebauman what about extracting values that contain `{` and `}` characters? Do we need tests for that? Any other weird limitations?

stevebauman committed:
Adjust formatting and lookup placeholders first

stevebauman committed:
Add more edge case testing with failing case

stevebauman committed:
Add result to failing test

stevebauman commented:
@taylorotwell I added more edge case tests and was able to break it here:

https://github.com/laravel/framework/pull/53781/files#diff-be47dbd336aece128f999b4c2d8b4fc7ee809a1d1e0ddd178547542201d4e897R254

```php
$values = Str::extract('{\\{bar}\\}', '*{foo}\\}');

// ['foo' => '}']
var_dump($values);

// Should be:
// ['foo' => '{bar}']
```

Not sure if this okay to have, let me know! Maybe a regex wizard and drop in and lend a hand 🧙

jasonmccreary commented:
I don't know about _wizard_, but one approach would be to tighten the "placeholder" values to be `{name}` where `name` is any valid PHP variable name. I'll take a quick look to see if I can get these passing...

jasonmccreary commented:
Took a quick look. I'm glad to look more, but the root problem here is going to be the root problem with any dynamic regular expression - escaping. Specifically what needs to be escaped versus what is already escaped.

The example you have is rather dense and focuses on braces. But the wildcard presents the same issues. Consider:

```php
Str::extract('what about *bold* words', '*\*{word}\**'));
```

I'm not sure if this is an intended use, but that's what I would expect to write when trying to _extract_ the word `bold`.

Similar to Blade parsing (which is done with regular expression), there are limits to their use. Those have to be mitigated with rules. For example, when using `Str:extract()`, you only need to escape literal instances of `*` or _placeholders_ (not _loose_ braces).

stevebauman committed:
Simplify pattern

stevebauman committed:
Fix test case

stevebauman committed:
Add more tests

stevebauman commented:
Thanks @jasonmccreary! Really appreciate you diving into this 🙏 

Yea you're right. We're going to hit some limitations and edge-cases here regardless due to regex's limitations.

I think my previously provided example wasn't a display of a good edge case, as we can actually just escape the brackets in the pattern (which is what we can always control) vs what is provided as a haystack string (which isn't what we can always control):

```php
// ['foo' => '{bar}']
Str::extract('{{\\{bar}\\}}', '*{\{foo}\}*');
```

I've added a bunch more edge case assertions and simplified the pattern a bit.

I think it's solid now. Let me know your thoughts!

CamKem commented:
> I've added a bunch more edge case assertions and simplified the pattern a bit.
> 
> I think it's solid now. Let me know your thoughts!

@stevebauman - Looks good, I think the only way to really improve on the reliability would be to tokenize it, however thats probably far to much complexity for what you're trying to achieve.

I put it through it's paces, and only escape chars that are really causing the issues (as you're aware), for example this:
```php
$pattern = '*/{first}/{second}/*';
$haystack = '123/131415*\\/\\d+d/7\\8\\9/101112';
$expected = [
    'first' => '131415*\\',
    'second' => '\\d+d',
];
//Expected: 131415*\, got: \d+d
//Expected: \d+d, got: 7\8\9
```

Setting more explicit / clear placeholder boundaries would allow this pattern to pass as expected, and could be a better option? However, doesn't help with Jasons edge case.
```php
    public static function extract($haystack, $pattern)
    {
        $placeholders = static::matchAll('/\{([^{}]+)}/', $pattern);
        
        $pattern = preg_quote($pattern, '/');

        foreach ($placeholders as $placeholder) {
            $pattern = static::replace(
                preg_quote('{'.$placeholder.'}', '/'),
                '(?<'.$placeholder.'>[^\/]+?)',
                $pattern,
            );
        }

        $pattern = static::replace('\*', '.*?', $pattern);
        $pattern = static::replace('\{', '{', $pattern);
        $pattern = static::replace('\}', '}', $pattern);

        if (preg_match("/^{$pattern}$/i", $haystack, $matches)) {
            return array_intersect_key($matches, $placeholders->flip()->all());
        }

        return [];
    }
```

I think it will be a great feature so long as the limitations & requirements are documented, especially as I can see people who are not comfortable with Regex reaching for it. 

Should ensure the docs cover:
- Placeholder names should be unique, made up of letters, numbers and underscores (no reserved characters)
- Reserved characters can't be used as delimiters: (\, *, ^, !, @).
- The effects of backslashes or delimiters appearing in the value to be extracted when parsing user input data.

jasonmccreary commented:
I agree with the tokenization. But I believe this could be achieved with the understanding you only escape the special characters used by `extract` (i.e. `*`, `{`, `}`).

If you all want to beef up the test cases with real-world examples, I'll get them passing.

stevebauman committed:
Small refactor

Thanks to @CamKem

stevebauman commented:
Thanks @CamKem! I've updated the PR with your suggestions and edge case. Appreciate you helping with this 🙏 

@jasonmccreary There's quite a few assertions already (34 atm) -- do you think we need more?

jasonmccreary commented:
I still think there are some oddities here. Some of these "edge cases" don't seem very real-world and/or break what I would consider the natural escaping rules for `extract`.

For example, the same result of these two is odd.

```php
$this->assertSame(['foo' => '{bar}'], Str::extract('{\\{bar}\\}', '*{\{foo}\}*'));
$this->assertSame(['foo' => '{bar}'], Str::extract('foo bar* {{bar}}', '* {{foo}}'));
```

The first assertion seems to _escape_ the placeholder. Furthermore, it escapes the inner opening brace, but the outer closing brace. This either (incorrectly) implies the name can contain braces or reveals a loophole in the escaping. The second assertion escapes nothing. Which may be fine. But odd that both are true.

I think using a real world example instead of simply trying to break the algorithm helps reveal the escaping problem. Consider wanting to extract digits from a partially masked credit card number. I'd personally write the following: But this doesn't work.

```php
$this->assertSame(
    ['first_4' => '1234', 'last_4' => '7890'], 
    Str::extract('1234*********7890', '{first_4}*\*{last_4}')
);
```

I don't mean to noisy up this PR. I think this is a cool feature. Would just advocate for it to follow the same rules/ideas as its underlying regular expressions. Or, at the least, have a clear rule of when/what to escape.

jasonmccreary commented:
Another set of real-world examples that fail. These reveal issues with the "greediness" of the underlying regex, as well as what it matches (e.g. `[^\/]+?`)

```php
$this->assertSame(
    ['first_4' => '4242'],
    Str::extract('4242-4242-4242-5000', '{first_4}*')
);
$this->assertSame(
    ['path' => 'docs/11.x/strings'],
    Str::extract('https://laravel.com/docs/11.x/strings#method-str-is', 'https://laravel.com/{path}#*')
);
```

Again, all solveable with defining some rules around how this is used. But felt these should be noted as they are different issues than the escaping previously noted.

taylorotwell commented:
Kinda fear this could be a black hole of edge cases and additional PRs, so going to table this one for now. 😅

stevebauman commented:
@taylorotwell haha totally fair! Thanks for your time 🙏
